File: swiftui/Sources/ArenaTokens/components/ArenaButton.swift
import SwiftUI

public enum ArenaButtonPaint {
    public static func fill(_ variant: ArenaButtonVariant, _ colors: ArenaColorScheme) -> Color {
        switch variant {
        case .primary: colors.accent
        case .secondary: colors.surfaceCard
        case .ghost: Color.clear
        case .danger: Color.clear
        }
    }

    public static func ink(_ variant: ArenaButtonVariant, _ colors: ArenaColorScheme) -> Color {
        switch variant {
        case .primary: colors.onAccent
        case .secondary: colors.textStrong
        case .ghost: colors.textStrong
        case .danger: colors.danger
        }
    }

    public static func edge(_ variant: ArenaButtonVariant, _ colors: ArenaColorScheme) -> Color {
        switch variant {
        case .primary: colors.accent
        case .secondary: colors.borderStrong
        case .ghost: Color.clear
        case .danger: colors.danger
        }
    }
}

public struct ArenaButton: View {
    @Environment(\.arenaColors) private var colors
    @Environment(\.arenaDensity) private var density
    @Environment(\.arenaFonts) private var fonts
    @FocusState private var focused: Bool

    private let content: String
    private let click: () -> Void
    private let variant: ArenaButtonVariant
    private let size: ArenaControlSize
    private let disabled: Bool
    private let loading: Bool
    private let full: Bool
    private let increasedContrast: Bool
    private let targetFloor: CGFloat?

    public init(
        _ content: String,
        variant: ArenaButtonVariant = .primary,
        size: ArenaControlSize = .md,
        disabled: Bool = false,
        loading: Bool = false,
        full: Bool = false,
        increasedContrast: Bool = false,
        targetFloor: CGFloat? = nil,
        click: @escaping () -> Void
    ) {
        self.content = content
        self.variant = variant
        self.size = size
        self.disabled = disabled
        self.loading = loading
        self.full = full
        self.increasedContrast = increasedContrast
        self.targetFloor = targetFloor
        self.click = click
    }

    private var inactive: Bool { disabled || loading }

    private var painted: CGFloat { ArenaScale.control(ArenaControl.height(size, density)) }

    private var box: CGFloat {
        guard let floor = targetFloor else { return painted }
        return ArenaControl.target(painted, floor: floor)
    }

    private var shape: RoundedRectangle { RoundedRectangle(cornerRadius: ArenaControl.radius) }

    private var ring: CGFloat { ArenaContrast.focusWidth(increasedContrast) }

    private var gutter: CGFloat { ArenaTokens.focusOffset + ring }

    private var ringShape: RoundedRectangle { RoundedRectangle(cornerRadius: ArenaControl.radius + gutter) }

    public var body: some View {
        Button(action: click) {
            HStack(spacing: ArenaControl.gap) {
                Text(content)
                    .font(fonts.body.font(size: ArenaScale.text(ArenaControl.text(size, density))))
                    .fontWeight(ArenaControl.weight)
                    .foregroundStyle(ArenaButtonPaint.ink(variant, colors))
            }
            .frame(maxWidth: full ? .infinity : nil, minHeight: box)
            .padding(.horizontal, ArenaControl.padding)
            .background(shape.fill(ArenaButtonPaint.fill(variant, colors)))
            .overlay(shape.strokeBorder(ArenaButtonPaint.edge(variant, colors), lineWidth: ArenaContrast.border(increasedContrast)))
            .contentShape(shape)
        }
        .buttonStyle(.plain)
        .focused($focused)
        .padding(gutter)
        .overlay(ringShape.strokeBorder(focused ? colors.focusRing : Color.clear, lineWidth: ring))
        .disabled(inactive)
        .accessibilityLabel(content)
        .accessibilityAddTraits(.isButton)
        .accessibilityRemoveTraits(inactive ? .isButton : [])
        .accessibilityAction { click() }
    }
}

File: compose/src/main/kotlin/org/dravensoft/arena/components/ArenaButton.kt
package org.dravensoft.arena.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.sizeIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.onClick
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import org.dravensoft.arena.api.ArenaButtonVariant
import org.dravensoft.arena.api.ArenaControlSize
import org.dravensoft.arena.theme.ArenaTheme
import org.dravensoft.arena.tokens.ArenaColorScheme
import org.dravensoft.arena.tokens.ArenaContrast
import org.dravensoft.arena.tokens.ArenaControl
import org.dravensoft.arena.tokens.ArenaScale
import org.dravensoft.arena.tokens.ArenaTokens
import org.dravensoft.arena.tokens.accent
import org.dravensoft.arena.tokens.borderStrong
import org.dravensoft.arena.tokens.danger
import org.dravensoft.arena.tokens.focusRing
import org.dravensoft.arena.tokens.onAccent
import org.dravensoft.arena.tokens.surfaceCard
import org.dravensoft.arena.tokens.textStrong

public object ArenaButtonPaint {
    public fun fill(variant: ArenaButtonVariant, colors: ArenaColorScheme): Color = when (variant) {
        ArenaButtonVariant.Primary -> colors.accent
        ArenaButtonVariant.Secondary -> colors.surfaceCard
        ArenaButtonVariant.Ghost -> Color.Transparent
        ArenaButtonVariant.Danger -> Color.Transparent
    }

    public fun ink(variant: ArenaButtonVariant, colors: ArenaColorScheme): Color = when (variant) {
        ArenaButtonVariant.Primary -> colors.onAccent
        ArenaButtonVariant.Secondary -> colors.textStrong
        ArenaButtonVariant.Ghost -> colors.textStrong
        ArenaButtonVariant.Danger -> colors.danger
    }

    public fun edge(variant: ArenaButtonVariant, colors: ArenaColorScheme): Color = when (variant) {
        ArenaButtonVariant.Primary -> colors.accent
        ArenaButtonVariant.Secondary -> colors.borderStrong
        ArenaButtonVariant.Ghost -> Color.Transparent
        ArenaButtonVariant.Danger -> colors.danger
    }
}

@Composable
public fun ArenaButton(
    content: String,
    click: () -> Unit,
    modifier: Modifier = Modifier,
    variant: ArenaButtonVariant = ArenaButtonVariant.Primary,
    size: ArenaControlSize = ArenaControlSize.Md,
    disabled: Boolean = false,
    loading: Boolean = false,
    full: Boolean = false,
    increasedContrast: Boolean = false,
    targetFloor: Dp? = null,
) {
    val colors = ArenaTheme.colors
    val density = ArenaTheme.density
    val inactive = disabled || loading
    val painted = ArenaScale.control(ArenaControl.height(size, density), LocalDensity.current.fontScale)
    val shape = RoundedCornerShape(ArenaControl.radius)
    val ring = ArenaContrast.focusWidth(increasedContrast)
    val gutter = ArenaTokens.focusOffset + ring
    var focused by remember { mutableStateOf(false) }
    Row(
        modifier = modifier
            .then(if (full) Modifier.fillMaxWidth() else Modifier)
            .onFocusChanged { focused = it.isFocused }
            .border(ring, if (focused) colors.focusRing else Color.Transparent, RoundedCornerShape(ArenaControl.radius + gutter))
            .padding(gutter)
            .sizeIn(minHeight = if (targetFloor == null) painted else ArenaControl.target(painted, targetFloor))
            .clip(shape)
            .background(ArenaButtonPaint.fill(variant, colors))
            .border(ArenaContrast.border(increasedContrast), ArenaButtonPaint.edge(variant, colors), shape)
            .clickable(enabled = !inactive, onClick = click)
            .padding(horizontal = ArenaControl.padding)
            .semantics {
                role = Role.Button
                contentDescription = content
                if (inactive) disabled()
                onClick { click(); true }
            },
        horizontalArrangement = Arrangement.spacedBy(ArenaControl.gap, Alignment.CenterHorizontally),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        BasicText(
            text = content,
            style = TextStyle(
                color = ArenaButtonPaint.ink(variant, colors),
                fontSize = ArenaControl.text(size, density),
                fontWeight = ArenaControl.weight,
                fontFamily = ArenaTheme.fonts.body,
            ),
        )
    }
}
